### ISSUE TITLE
[FEATURE] .megaignore file

### ISSUE THREAD

**Issue by curz46** [meganz/MEGAsync#206]
Hi,
I'm aware that MEGA has a pattern-matching ignore config that can be accessed through the GUI, but I think that a `.megaignore` file would still be useful:
- much easier to manage and edit than the GUI.
- can be uploaded to (MEGA) remote, so that you don't have to configure sync ignore on every device.
- allows users running MEGA in headless mode to ignore files and folders.
- enables `!<pattern>` syntax; electing to sync certain files that would otherwise be caught by ignore config.
- massively requested feature from DB community.

It would run in parallel to the current GUI-based ignore, overriding it where `!` syntax is used. 

Questions:
- is something like this already implemented? 
- how plausible would this be to implement?
- would a PR implementing this be accepted?
- where would it best fit into the application?

Thanks, Dylan

**Comment by mohamedhk2** [comment]
+1

**Comment by rayhanethk2** [comment]
+1

**Comment by KsibiMohamed** [comment]
+1

**Comment by haithemkdous** [comment]
+1

**Comment by DevPre24** [comment]
+1

**Comment by davidsonalencar** [comment]
+1

**Comment by Sh3Rm4n** [comment]
+1

**Comment by salaros** [comment]
I was going to create a feature request just like this, Advanced settings tab is great, but I'm using dual-boot on my system, so I have to keep in mind to keep exclusion lists even on both clients on Windows and Linux.
Additionally I've recently bought a new laptop and just switched plugged in secondary HDD to the new PC. I forgot to copy the exclusion list to the newly installed MEGAsync client, so it has uploaded unwanted content to the cloud. 

With .megaignore file feature this wouldn't have happened + I wouldn't have to keep Linux and Windows client's settings even

+1

**Comment by ankurdubey26** [comment]
+1

**Comment by caiocampos** [comment]
+1

**Comment by abwesend890** [comment]
+1

**Comment by furioness** [comment]
Please. When creating projects using npm with many little files in many folders - megasync becomes mad.

**Comment by komidawi** [comment]
+1

**Comment by leobel96** [comment]
Any news?

**Comment by curz46** [comment]
Haven't heard anything from them. I think it's probably very low priority for their team and doesn't seem like they're accepting contribution

**Comment by dwmega** [comment]
Test

**Comment by iinfin** [comment]
Still waiting I guess, it'd be a really useful feature

**Comment by JoshuaDietz** [comment]
I'm waiting too. I don't understand why such an important, relatively small and much requested feature is not implemented even after almost two years.

**Comment by voxsoftware** [comment]
Still waiting this feature. I use mega-cmd. I need .megaignore file or something similar

**Comment by valerii540** [comment]
+1

**Comment by nukadelic** [comment]
+1

**Comment by gahfaustino** [comment]
+1
When I'm developing certains app than have package folders like npm node_modules, megasync going crazy with thousands and thousands files.

This feature gonna be awesome. While this not happen I will need stop using mega for this projects.

**Comment by sguarin** [comment]
+1

**Comment by nuno1212s** [comment]
+1

**Comment by nukadelic** [comment]
plus one

**Comment by Co-Verse** [comment]
welp

**Comment by arekscode** [comment]
+1 
Similar issue as the aforementioned NPM problem. I use mega to sync a Firefox profile across devices and there are a lot of temporary files that Firefox creates which don't need to get synced.

**Comment by nukadelic** [comment]
+1
bumping it up [ again ] 
<3 mega

here is a motivational duck ( audio ) : 

https://user-images.githubusercontent.com/6582633/119255325-c7489c80-bbed-11eb-82ec-055d856da05b.mp4

**Comment by Lilac-Orange** [comment]
+1

**Comment by fstnando** [comment]
+1

**Comment by salaros** [comment]
+2

**Comment by WeaverOfTheWeb** [comment]
Hello, My great great grandfather once spoke of this feature and here we are, still waiting on it...

**Comment by BrunoMont2003** [comment]
Still waiting .... -_-

**Comment by mattw-mega** [comment]
Hi to you keen people.

We have been working on this.  I would like to share our plans for how the .megaignore file will work, and it would be great to get any feedback - will it cover your cases, do you see any problems, etc.  thanks

.megaignore File Format

Ignore files are flat text files named '.megaignore' that control which files or directories are synchronized.

The most common use is with one .megaignore file in the sync root with its rules deciding inclusion and exclusion for each directory and file in the whole sync. 

When a sync is started, a default .megaignore file will be added in the sync root folder, if one does not exist yet.   

Syncs only load and apply the rules for a .megaignore file once it is present in the local filesystem side of the sync.   If there is a .megaignore file on the cloud side, but not the local side, that file will first be synced, and its rules loaded, before the subtree starts syncing.   Other than that, .megaignore files are synced in a similar fashion to other files, and it's only once the file arrives locally that its rules apply, so it's best to ensure adjustments to those are fully synced before making other changes that need those new rules in place.

Users can also add .megaignore files in any folder within the sync, to add inclusion and exclusion rules for that subtree of the sync.  The files can also be edited locally, and those changes will be applied when the directory is next processed.   Moving or renaming .megaignore files can mean that the rules from those files are suddenly not present, so caution is advised there.

If a .megaignore file contains errors, such as not following the syntax below, or due to one or more of the contained rules not being understandable when loaded into the sync, then that folder cannot be synced, and a sync stall will result.   Other parts of the sync may be affected too, as it may not be possible to determine which folders to scan and whether missing items were moved or not.   The user can check the sync stall issues to see where these problems occurred, and then fix them.

Removal of a .megaignore file also removes its rules from the sync.

FILE FORMAT
Ignore files contain comments, whitespace or filters.

Comments are used for documentation purposes and are ignored when loading a .megaignore file. You start a comment by beginning a line with the '#' character. Note that the '#' character must not be preceded by any whitespace and that the comment ends at the end of the line.

Filters are specified line by line and come in two forms: Name filters and size filters.

NAME FILTERS
Name filters are matched against a file's name or type. You can use these kinds of filters to exclude directories or files whose name matches a specific pattern.

Name filters have the general format:

<CLASS><TARGET><TYPE><STRATEGY>:<PATTERN>

<CLASS>
Filters must be one of the following two classes: 'exclude' or 'include'.

- : Filters of the 'exclude' class specify files or directories that should not be synchronized.
This class is indicated by the '-' character.

+ : Filters of the 'include' class specify files or directories that should be synchronized.
This class is indicated by the '+' character.

<TARGET>
Filters may have one of the following three targets: 'all', 'directory' or 'file'.

a : The 'all' target specifies that this filter applies to both files and directories and is indicated by the 'a' character.

d : The 'directory' target specifies that this filter applies only to directories and is indicated by the 'd' character.

f : The 'file' target specifies that this filter applies only to files and is indicated by the 'f' character.

The a  'all' target will be used if you do not explicitly specify a target.

<TYPE>
Filters may be one of the following three types: 'local name', 'subtree name' or 'path'.

N : 'local name' filters are indicated by the 'N' character and are matched against a file or directory's name. Unlike 'subtree name' filters, they have an effect only in the the directory containing the .megaignore file.

n : 'subtree name' filters are indicated by the 'n' character and are matched against a file or directory's name. Unlike 'local name' filters, they have an effect in the directory containing the .megaignore file as well as its children.

p : 'path' filters are indicated by the 'p' character and are matched against the file or directory's path relative to the directory containing the .megaignore file. Note that the path separator is always '/', even on Windows.

The n 'subtree name' type will be used if you do not specify a type explicitly.

<STRATEGY>
Filters may use one of the following two match strategies: 'glob' or 'regexp'.

Gg : 'glob' filters are indicated by the 'g' or 'G' characters and match against a name or path using a 'wildcard pattern'.

Rr : 'regexp' filters are indicated by the 'r' or 'R' characters and match against a name or path using a pattern expressed as a POSIX Extended Regular Expression.

The 'glob' match strategy will be used if you do not explicitly specify a match strategy.

Uppercase ('G' or 'R') specifies that matching should be case sensitive.

<EXAMPLES>

# This is a comment.
# Everything after and including the # is ignored.

# This is a subtree name exclusion matched using a wildcard pattern.
#
# It will exclude all *.txt files in and beneath the directory containing
# the .megaignore file.
-f:*.txt

# This is a subtree name inclusion matched using a wildcard pattern.
# It will include all work*.txt files excluded by the *.txt exclusion
# above.
+fg:work*.txt

# This is also a subtree name exclusion matched using a wildcard pattern.
# It will exclude all *.jpg files in and beneath this directory.
-fn:*.jpg

# This is also a subtree name inclusion matched using a wildcard pattern.
# It will include all family*.jpg files previously excluded by the *.jpg
# exclusion above.
+fng:family*.jpg

# This is a local name exclusion matched using a wildcard pattern.
# It will exclude all *.avi files contained directly by this directory.
-N:*.avi

# This is a path exclusion matched using a wildcard pattern.
# It will exclude all *.avi files in the video subdirectory.
-fp:video/*.avi

# This is a path inclusion matched using a wildcard pattern.
# It will include all *.avi files in the video/family subdirectory.
+fpg:video/family/*.avi

# This is a subtree name exclusion matched using a regular expression.
# It will exclude all files whose name contains foo.
-nr:.*foo.*

# This will exclude all directories with the name 'private'
-d:private

# Exclude everything in and below this directory.
-:*

SIZE FILTERS

Size filters match against a file's size and can be used to exclude files whose size is above, below or between a specified threshold.

Size filters are only intended to avoid adding files to the sync.  Once a file is synced, size filters have no further effect.  Therefore, adjusting the size thresholds won't stop already synced files continuing to be synced, and neither will edits of files that result in a different size.

Size filters have the general form:

<THRESHOLD>:<VALUE><UNIT>

<THRESHOLD>

The threshold element specifies which bound you're specifying a value for.

There are two thresholds that you can specify:

exclude-smaller : Specifies the lower-bound. Files smaller than the specified value will be excluded.
exclude-larger : Specifies the upper bound. Files larger than the specified value will be excluded.

If you specify both thresholds, files with a size outside of the specified range will be excluded.

It is invalid to specify a lower bound greater than (or equal to) the upper bound.

<VALUE>

The value element states the bound for a specified threshold.

Any non-negative integer value is valid.

<UNIT>

The optional unit element specifies which unit the value is stated in.

K : The value is specified in kilobytes.
M : The value is specified in megabytes.
G : The value is specified in gigabytes.

Units are case insensitive so "k" would also specify "kilobytes."

If no unit is specified, the value is interpreted as being in bytes.

<EXAMPLES>

# Exclude files smaller than 4 kilobytes.
exclude-smaller:4k

# Exclude files larger than 8 megabytes.
exclude-larger:8M

# Exclude files smaller than 4 kilobytes and larger than 8 megabytes.
exclude-smaller:4K
exclude-larger:8m

CAVEATS

Note that when a .megaignore file is being downloaded, its rules do not come into effect until that download has completed and the .megaignore file has been successfully loaded. It's very important to keep this in mind if you try to move .megaignore files around in the cloud.

An example of how this behavior might surprise you, consider the case of moving a .megaignore file up a directory hierarchy. You might expect that the .megaignore file's rules remain in effect at the source directory until the .megaignore file appears at its new location.

This is not necessarily the case as the sync engine may interpret these events as the .megaignore file being removed at the source and a new .megaignore file being added at the target. In this case, the rules present at the source are purged and are only reapplied when the .megaignore file has been redownloaded.

If you're moving .megaignore files around and it's important that the filters stay in effect at the source until the move has completed, you will want to copy the .megaignore file to the new location first then remove the original after the copy has been synchronized.

**Comment by eddieparker** [comment]
Thanks Mattw!  Love that you're working on this.

Some feedback:

- Love the idea of having multiple .megaignores and that they apply inclusion/exclusion rules to the subtree.
- My gut instinct says that the megaignore should act as close to a .gitignore as possible to make onramping easy.  But maybe you have good reasons to diverge from this pattern?
- It's a bit confusing having the file named .megaignore but it does both exclusion AND inclusion?  Being named .megaignore feels like it should only ignore.  Perhaps .megafilters would be a more indicative name?
- The 'target' syntax seems a bit heavy handed.  Couldn't you just use a regex or something?  i.e. Files could be ".*/foo.txt$" vs a path would be "/my_dir_name/", whereas both would just be a "<my_file_or_path_name>"?  
- Similarly the 'type' syntax.  .gitignore seems to have a lot of this baked with using "**" and the like, similarly the above regex idea could be versatile in describing how far down the tree it goes.
- The size filter stuff is neat, I hadn't thought about that.
- Am I right in undersatnding that the only time you need to use '+' filters is to "undo" part of a '-' filter?  Makes sense if so; I'm assuming "everything" is "+" by default effectively.

Just another idea, in case the team does go down this route of explicitly specifying things - perhaps .yaml might be a better file format so it self documents a bit better.  I can foresee already needing to copy/paste the 'man' page you have here into my .megaignore file or what have you so I can remember what '-fpg:' is vs '-nr'.  Versus having the ability to do something like:

```
- class: exclude
  type: path
  strategy: glob
  path: "video/family/*.avi"

- class: exclude
  type: subtree
  strategy: regex
  path: ".*foo.*"
```
  
Would make things a lot more readable and easier to copy-paste without issues.

Still, love that there's progress here!

**Comment by curz46** [comment]
This looks really good. I'm super stoked to see this feature request get some attention after quite some time! My thoughts:
1. I like that you've made the file format quite expressive in options provided, which overall makes MEGA a much more attractive filesync option.
2. If it's called .megaignore, it should probably match .gitignore syntax for intuitiveness. i.e. maybe just replace `+` with `!` and make `-` the default? I think it's fine that it can also "undo" changes with `+` entries like Eddie mentioned, because Git does the same thing with `!` (and this is super useful.)
3. For the `N`, `n`, `p`, and even `a`, `d`, `f` options, Git leans on glob. Is there a reason why you think explicit options would work better? However, I think this is not a deal breaker especially if it aids in efficiency. The important part is these all have sensible defaults, and I agree with what you've set. 
4. In the same spirit of (2.) I saw you showed no options would be listed as `-:some/path/to/include`. Maybe we can drop the `:` in this case for brevity, which along with (2.) would make it a compatible extension of `.gitignore` syntax. I could even `cp .gitignore .megaignore` if that's something I would want to do per my use case.
5. The `exclude-smaller` and `exclude-larger` options seem heavy-handed and out-of-place as compared to the rest of the spec, maybe `>` and `<` respectively, parsed in the same way as `-` and `+` would work better? If you want to make it explicit that we're excluding here (though I think this is heavily implied, since you would not want to "undo" for very massive files), you could mandate `:` (like you already planned) and use syntax like `->:` and `-<:`.

Thanks a lot for the work so far, even this current spec. seems great (I am nitpicking really.) Good job!

**Comment by JimTheKiwifruit** [comment]
So glad to see this is being worked on 🎉 I'm 100% switching to MEGA as my primary cloud storage when this is implemented. I agree with @curz46 point 2, mainly having exclusion as the default like gitignore (as for me that will be 99% of what I would use it for).

**Comment by mattw-mega** [comment]
Hi to everyone who posted in this thread.   We are close to being ready to release this .megaignore feature.   It's part of a very large set of changes for MEGAsync, so we would like to invite you all to try it out and give us any feedback or let us know if you encounter any bugs:  https://github.com/meganz/MEGAsync/issues/770

Thanks especially to those who posted feedback about the .megaignore rules.   We considered it but have not really changed anything, we think the format is quite good for the general public.  If you're a developer using git, you are certainly smart enough to figure out how to use these :-).  About the exclude-larger and exclude-smaller, yes they deliberately stand out in order to be very clear whether they are including or excluding.   Also they are quite different in that it's possible to have files included according to the cloud side size, but excluded according to the local side size, which can't happen for the rules based on name.

Just tagging everyone in case that helps alert you to this:  @JimTheKiwifruit @webdev69420 @curz46 @eddieparker @BrunoMont2003 @WeaverOfTheWeb @salaros @Lilac-Orange @nukadelic @arekscode @Knosiz @valerii540 @voxsoftware @JoshuaDietz @iinfin @leobel96 @komidawi @furioness @RagedPanda @mohamedhk2

**Comment by salaros** [comment]
Offtopic: I waited for ignore file feature for so long (on MEGAsync and later OneDrive) and had to switch to another cloud provider which HAD this feature,
even though I was a paying MEGA customer for few years and still have an active Office 365 subscription

**Comment by thiagovilla** [comment]
So. Is it out yet?

**Comment by maticboncina** [comment]
This out yet?

**Comment by eddieparker** [comment]
@thiagovilla / @maticboncina : mattw linked tot he alpha client you can use to test it out.  Sounds like it's not out in the general release, I'm assuming they'll close this bug when it is.  In the meantime, try out the beta and give them feedback.

**Comment by Lautaro-De-Lucia** [comment]
Any deadline for this to be part of the official release? Can't tell you how much I need this feature

**Comment by mattw-mega** [comment]
Hi All, a new refinements have been made, which you can try out at #770 (alpha software so please be careful)

 - .megaignore will not itself be synced by default.   You can make it be synced by adding the rule `+sync:.megaignore` to the file itself.   Turn the + to a - to reverse the rule (or leave the line out).
 - this will make the upgrade from the old GUI based rule set smoother for those with multiple devices syncing the same folder, allowing for different rule sets per device.  But also giving you the choice of syncing those rules if you want to, or keep the flexibility of different rule sets per device.
 - it will also support the possibility of avoiding syncing large subfolders on devices without much storage.  Eg. the root level .megaignore might be synced, but a particular subfolder might have a non-synced one.
 - Your old GUI based rules will additionally be migrated to a .megaignore.default file in the app's data/db folder.   You can keep the rules in that one up to date, and new syncs will get a copy of that rule set.

thanks

**Comment by gherbetto** [comment]
it's here guys!!! finally ignore file for mega in version 5.0.22

**Comment by jbhardman** [comment]
Is it ever going to be possible to use .megaignore with mega-cmd sync?

**Comment by gherbetto** [comment]
> > it's here guys!!! finally ignore file for mega in version 5.0.22
> 
> Awesome! How can we update?

sorry i didn't notice your question. mega emailed us about this proposing to try this feature. it was available by following a link with the download. upon search there is a mac and windows option only.

**Comment by cihe13375** [comment]
> sorry i didn't notice your question. mega emailed us about this proposing to try this feature. it was available by following a link with the download. upon search there is a mac and windows option only.

If anyone wonders, currently it's still possible to download the linux build of v5.0.22 alpha at here https://mega.nz/folder/xs5DVbZY#gnVsjU6AchUYrBZornsOuQ (if you use windows or mac go for mega homepage to download the official release of v5.2 instead).

*Disclaimer: alpha version is alpha version, so bugs are expected. Personally I have installed it on my linux machine for months and generally it works fine, but your experience may vary. I am not mega stuff so I cannot provide much support, and cannot be responsible for any consequences caused by using it. Install at your own risk (you may want to browse the thread of alpha testing https://github.com/meganz/MEGAsync/issues/770 first).*

**Comment by tomjtoth** [comment]
anyone having luck with re-inclusion of select files after a `-:*` ?
I mean I'd like to sync about 15 items of 30 items in a folder without setting up 15 syncs, here's my `.megaignore` content:

```
+sync:.megaignore
-N:*
# all symlinks
-s:*

-d:__pycache__
-d:venv
-d:target
-d:.git

# bash related
+N:.{bashrc,bash_{aliases,logout,profile},profile}

# etc
+N:.{android,gitconfig,ssh}

+N:{bin,Documents,Education,Games,Music,Pictures,Projects,Videos}
+p:.config/{htop,mc,qBittorrent,sway,systemd}
```

**Comment by gherbetto** [comment]
> v5.2 linux official is released. I guess this issue can be resolved now...

Is this available on megacmd with the latest Linux update?

**Comment by webdev69420** [comment]
This is finally a feature, thanks guys!

**Comment by cperonneaud** [comment]
> Hi All, new refinements have been made, which you can try out at #770 (alpha software so please be careful)
> 
> * .megaignore will not itself be synced by default.   You can make it be synced by adding the rule `+sync:.megaignore` to the file itself.   Turn the + to a - to reverse the rule (or leave the line out).
> * this will make the upgrade from the old GUI based rule set smoother for those with multiple devices syncing the same folder, allowing for different rule sets per device.  But also giving you the choice of syncing those rules if you want to, or keep the flexibility of different rule sets per device.
> * it will also support the possibility of avoiding syncing large subfolders on devices without much storage.  Eg. the root level .megaignore might be synced, but a particular subfolder might have a non-synced one.
> * Your old GUI based rules will additionally be migrated to a .megaignore.default file in the app's data/db folder.   You can keep the rules in that one up to date, and new syncs will get a copy of that rule set.
> 
> thanks

I'm having an issue with the .megaignore being ignored for some reason. I've a Unity project with ~14GB of data and ~31Gb of Unity garbage. I didn't know that when I first synced and now, I have 45Gb of data on Mega. I tried to add the garbage folder to the megaignore list but it seems to be ignored and it's still syncing files inside the folder. I tried to delete the folder on my Mega and now, MegaSync says there's a conflict... I'm a bit lost... Why are my ignore rules ignored by MegaSync?

**Comment by gherbetto** [comment]
> > Hi All, new refinements have been made, which you can try out at #770 (alpha software so please be careful)
> > 
> > * .megaignore will not itself be synced by default.   You can make it be synced by adding the rule `+sync:.megaignore` to the file itself.   Turn the + to a - to reverse the rule (or leave the line out).
> > * this will make the upgrade from the old GUI based rule set smoother for those with multiple devices syncing the same folder, allowing for different rule sets per device.  But also giving you the choice of syncing those rules if you want to, or keep the flexibility of different rule sets per device.
> > * it will also support the possibility of avoiding syncing large subfolders on devices without much storage.  Eg. the root level .megaignore might be synced, but a particular subfolder might have a non-synced one.
> > * Your old GUI based rules will additionally be migrated to a .megaignore.default file in the app's data/db folder.   You can keep the rules in that one up to date, and new syncs will get a copy of that rule set.
> > 
> > thanks
> 
> I'm having an issue with the .megaignore being ignored for some reason. I've a Unity project with ~14GB of data and ~31Gb of Unity garbage. I didn't know that when I first synced and now, I have 45Gb of data on Mega. I tried to add the garbage folder to the megaignore list but it seems to be ignored and it's still syncing files inside the folder. I tried to delete the folder on my Mega and now, MegaSync says there's a conflict... I'm a bit lost... Why are my ignore rules ignored by MegaSync?

We don’t really know what you mediaignore looks like to know how to help you.

**Comment by cperonneaud** [comment]
> We don’t really know what you mediaignore looks like to know how to help you.

Well, it's a pretty simple megaignore.

-:Thumbs.db
-:desktop.ini
-:~*
-:.*
-:*~.*
-:*.crdownload
-:*.sb-????????-??????
-:*.tmp
-:Glassix 2\Library\*
-d:Glassix 2\Library
-d:Glassix 2\Temp
-:Glassix 2/Temp

I'm backing up my game folder named Glassix 2 which contains my out of project files + Unity project. Inside this folder, there is another Glassix 2 folder which is the Unity project itself, and inside this Unity folder, there are two folders I don't want to include in the sync: Library and Temp. I've used MegaSync built in UI to add those two folders, I've also tried to edit the megaignore by adding the wildcard line but to no avail, for some reasons, the library and temp folders are still tagged with the little red x icon on Windows indicating it's not syncing properly and I also see some files from the Library being synced from time to time.
Another important thing was that this project was at first on another desktop at first which crashed. At that time, I was in a rush and instead of reinstalling MegaSync on the new desktop, I downloaded from Mega directly and started working on the game again without MegaSync. Now I'm trying to sync again and it does seem to work fine since it updated all the new files, just this Library and Temp folders which are too big and need to be removed.
One thing I noticed is that the status indicates "Syncing" in MegaSync and I'm wondering if it might require a full pass completion before taking the megaignore file into account maybe?

**Comment by tienll** [comment]
Can I ignore all hidden files by adding rule but exclude some of them? For example, I want to sync `.megaignore` to make sure the rules are same on all my devices, so I tried to add rules like this
```
+:.megaignore
-:.*
```
however it doesn't work for me.

**Comment by Farfi55** [comment]
## Introducing [git2megaignore](https://github.com/Farfi55/git2megaignore)

Hello everyone! Apologies in advance for the *self-promotion*, but I believe this tool could be particularly useful for followers of this discussion.

### What is it?

***git2megaignore*** is a Rust CLI tool that simplifies the creation of `.megaignore` rules by converting `.gitignore` rules into their `.megaignore` equivalents.

### How does it work?

At its core, git2megaignore processes the **glob patterns** in your `.gitignore` file and adapts them to work with the glob engine used by MegaSync. For example, it applies the appropriate MegaSync rule options, such as `d` for targeting *directories-only* when a rule ends with a `/`.

However, since MegaSync's glob engine doesn't support all `.gitignore` patterns, some rules must be converted into **Regular Expressions** instead. For instance, `/[Bb]uild/` will be translated into the RegEx `^[Bb]uild$`.

Here’s an example:

Your `.gitignore` file:
```gitignore
/[Tt]emp/
.vs/
*.tmp
.vscode/*
!.vscode/settings.json
```

git2megaignore will generate the following `.megaignore` rules:
```gitignore
-dpR:^[Tt]emp$
-dnG:.vs
-nG:*.tmp
-pG:.vscode/*
+pG:.vscode/settings.json
```

### Where can I get it?

If you're interested, check out the [git2megaignore repository](https://github.com/Farfi55/git2megaignore), where you’ll find more information about the project and installation instructions.

If you have questions about using the tool or encounter issues, feel free to open an issue on the repository. Feedback is greatly appreciated!

**Comment by Rashcodeacademy** [comment]
> ## Introducing [git2megaignore](https://github.com/Farfi55/git2megaignore)
> Hello everyone! Apologies in advance for the _self-promotion_, but I believe this tool could be particularly useful for followers of this discussion.
> 
> ### What is it?
> _**git2megaignore**_ is a Rust CLI tool that simplifies the creation of `.megaignore` rules by converting `.gitignore` rules into their `.megaignore` equivalents.
> 
> ### How does it work?
> At its core, git2megaignore processes the **glob patterns** in your `.gitignore` file and adapts them to work with the glob engine used by MegaSync. For example, it applies the appropriate MegaSync rule options, such as `d` for targeting _directories-only_ when a rule ends with a `/`.
> 
> However, since MegaSync's glob engine doesn't support all `.gitignore` patterns, some rules must be converted into **Regular Expressions** instead. For instance, `/[Bb]uild/` will be translated into the RegEx `^[Bb]uild$`.
> 
> Here’s an example:
> 
> Your `.gitignore` file:
> 
> ```gitignore
> /[Tt]emp/
> .vs/
> *.tmp
> .vscode/*
> !.vscode/settings.json
> ```
> 
> git2megaignore will generate the following `.megaignore` rules:
> 
> ```gitignore
> -dpR:^[Tt]emp$
> -dnG:.vs
> -nG:*.tmp
> -pG:.vscode/*
> +pG:.vscode/settings.json
> ```
> 
> ### Where can I get it?
> If you're interested, check out the [git2megaignore repository](https://github.com/Farfi55/git2megaignore), where you’ll find more information about the project and installation instructions.
> 
> If you have questions about using the tool or encounter issues, feel free to open an issue on the repository. Feedback is greatly appreciated!

Amazing Bro! Thanks. You are a great programmer.
